pkg/types/name.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"strings"
+
+	"github.com/iancoleman/strcase"
+)
+
+// NewNameFromSnake produces a Name, using given snake case string as source of
+// truth.
+func NewNameFromSnake(s string) Name {
+	originals := strings.Split(s, "_")
+	camels := make([]string, len(originals))
+	for i, org := range originals {
+		if known, ok := lowerToCamelAcronyms[strings.ToLower(org)]; ok {
+			camels[i] = known
+			continue
+		}
+		camels[i] = strcase.ToCamel(org)
+	}
+	return Name{
+		Snake:      s,
+		Camel:      strings.Join(camels, ""),
+		LowerCamel: strings.Join(append([]string{strings.ToLower(camels[0])}, camels[1:]...), ""),
+	}
+}
+
+// Name holds different variants of a name.
+type Name struct {
+	Snake      string
+	Camel      string
+	LowerCamel string
+}
+
+// Add acronyms that can be safely assumed to be common for any kind of provider.
+// For provider-specific ones, like ARN for Amazon Web Services, provider
+// authors need to configure them in their provider repository.
+
+// NOTE(muvaf): We can have more maps like camel -> lower for reverse conversion,
+// but it's not necessary for now.
+
+var (
+	// Used to hold lower -> camel known exceptions.
+	lowerToCamelAcronyms = map[string]string{
+		"id": "ID",
```

Comment:
Should we move this to`init()` or vice versa?

Reply:
I added that just as an example pairing so it's clear that the key is lower case and value is camel but I can move it to `init()` if you have a strong feeling.